File: src/main/java/UI/MainWindowController.java
package UI;

import javafx.beans.property.SimpleStringProperty;
import javafx.event.ActionEvent;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.AnchorPane;
import javafx.stage.Modality;
import javafx.stage.Stage;
import models.File;
import models.Package;
import models.User;
import services.FileService;
import services.PackageService;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.attribute.BasicFileAttributeView;
import java.nio.file.attribute.BasicFileAttributes;
import java.nio.file.attribute.FileTime;
import java.sql.Timestamp;
import java.util.*;

public class MainWindowController {

    private User user;

    void setUser(User user) {
        this.user = user;
        usernameLabel.setText("Files of " + user.getName());

        updateFilesTable();
        updatePackagesTable();
    }

    public AnchorPane rootAP;
    public Label usernameLabel;


    // files logic
    class FileRow {
        SimpleStringProperty nameProperty;
        SimpleStringProperty createdProperty;
        SimpleStringProperty lastModifiedProperty;

        FileRow(String name, Timestamp created, Timestamp lastModified) {
            this.nameProperty = new SimpleStringProperty(name);
            this.createdProperty = new SimpleStringProperty(created.toString());
            this.lastModifiedProperty = new SimpleStringProperty(lastModified.toString());
        }

        SimpleStringProperty getNameProperty() {
            return nameProperty;
        }

        SimpleStringProperty getCreatedProperty() {
            return createdProperty;
        }

        SimpleStringProperty getLastModifiedProperty() {
            return lastModifiedProperty;
        }
    }

    public Button addFileButton;
    public TreeTableView<FileRow> filesTree;
    public TreeTableColumn<FileRow, String> fileNameColumn;
    public TreeTableColumn<FileRow, String> createdColumn;
    public TreeTableColumn<FileRow, String> modifiedColumn;

    private void updateFilesTable() {
        FileService fileService = new FileService(user.getID(), user.getPassword());
        List<File> userFiles = fileService.getAll();

        TreeItem<FileRow> root = new TreeItem<>(new FileRow("n", new Timestamp(0), new Timestamp(0)));
        fileNameColumn.setCellValueFactory(
                (TreeTableColumn.CellDataFeatures<FileRow, String> param)
                        -> param.getValue().getValue().getNameProperty()
        );
        createdColumn.setCellValueFactory(
                (TreeTableColumn.CellDataFeatures<FileRow, String> param)
                        -> param.getValue().getValue().getCreatedProperty()
        );
        modifiedColumn.setCellValueFactory(
                (TreeTableColumn.CellDataFeatures<FileRow, String> param)
                        -> param.getValue().getValue().getLastModifiedProperty()
        );

        HashMap<String, ArrayList<FileRow>> filesByDir = new HashMap<>();
        for (File userFile : userFiles) {
            if (filesByDir.containsKey(userFile.getPath())) {
                filesByDir.get(userFile.getPath()).add(
                        new FileRow(userFile.getName(), userFile.getCreated(), userFile.getModified())
                );
            } else {
                filesByDir.put(
                        userFile.getPath(),
                        new ArrayList<>(
                                Collections.singletonList(
                                        new FileRow(userFile.getName(), userFile.getCreated(), userFile.getModified())
                                )
                        )
                );
            }
        }

        for (Map.Entry<String, ArrayList<FileRow>> entry : filesByDir.entrySet()) {
            TreeItem<FileRow> dir = new TreeItem<>(new FileRow(entry.getKey(), new Timestamp(0), new Timestamp(0)));
            for (FileRow _file : entry.getValue()) {
                TreeItem<FileRow> row = new TreeItem<>(_file);
                dir.getChildren().add(row);
            }
            root.getChildren().add(dir);
        }
        filesTree.setRoot(root);
        filesTree.setShowRoot(false);
    }

    public void addFileClicked() throws IOException {
        FXMLLoader loader = new FXMLLoader(getClass().getResource("/AddFileDialog.fxml"));
        Parent root = loader.load();
        AddFileDialogController controller = loader.getController();

        Stage stage = new Stage();
        stage.setScene(new Scene(root));
        stage.initModality(Modality.APPLICATION_MODAL);
        stage.setResizable(false);
        stage.showAndWait();

        String files_list = controller.filePathInput.getText();
        String[] files_path = files_list.split(",");

        for (String _path : files_path) {
            String path = _path.trim();
            java.io.File temp_file = new java.io.File(path);
            if (!temp_file.exists()) {
                if (!path.isEmpty()) {
                    Alert alert = new Alert(Alert.AlertType.ERROR);
                    alert.setTitle("Wrong file");
                    alert.setHeaderText("Wrong file");
                    alert.setContentText("File " + path + " does not exist!");
                    alert.showAndWait();
                }
                return;
            }

            if (temp_file.isFile()) {
                insertFileIntoDB(temp_file, path);
                updateFilesTable();
            }

            if (temp_file.isDirectory()) {
                insertCatalogIntoDB(temp_file);
                updateFilesTable();
            }
        }
    }

    private void insertFileIntoDB(java.io.File dirFile, String path) throws IOException {
        String name = path.substring(path.lastIndexOf('/')+1);
        String location = path.substring(0, path.lastIndexOf('/'));

        byte[] content = Files.readAllBytes(dirFile.toPath());

        File file = new File(
                name,
                location,
                content,
                Timestamp.from(
                        Files.readAttributes(
                                dirFile.toPath(),
                                BasicFileAttributes.class
                        ).creationTime().toInstant()
                ),
                Timestamp.from(
                        Files.readAttributes(
                                dirFile.toPath(),
                                BasicFileAttributes.class
                        ).lastModifiedTime().toInstant()
                )
        );

        FileService fileService = new FileService(user.getID(), user.getPassword());
        fileService.save(file);
    }

    private void insertCatalogIntoDB(java.io.File catalog) throws IOException {
        java.io.File[] filesInDir = catalog.listFiles();
        assert filesInDir != null;
        for (java.io.File dirFile : filesInDir) {
            if (dirFile.isDirectory()) {
                insertCatalogIntoDB(dirFile);
            } else if (dirFile.isFile()) {
                String path = dirFile.getAbsolutePath();
                insertFileIntoDB(dirFile, path);
            }
        }
    }

    private void downloadFile(File obj) throws IOException {
        String path = obj.getPath()+'/'+obj.getName();
        java.io.File file = new java.io.File(path);
        if (!file.exists()) {
            if (!file.getParentFile().mkdirs()) {
                return;
            }
            if (!file.createNewFile()) {
                return;
            }
        }
        Files.write(file.toPath(), obj.getContent());

        BasicFileAttributeView attributes = Files.getFileAttributeView(file.toPath(), BasicFileAttributeView.class);
        attributes.setTimes(
                FileTime.from(obj.getCreated().toInstant()),
                FileTime.from(obj.getModified().toInstant()),
                FileTime.from(obj.getModified().toInstant())
        );
    }


    // packages logic
    class PackageRow {
        SimpleStringProperty nameProperty;
        SimpleStringProperty repositoryProperty;
        SimpleStringProperty configsListProperty;

        PackageRow(String name, String repository, List<File> configsList) {
            this.nameProperty = new SimpleStringProperty(name);
            this.repositoryProperty = new SimpleStringProperty(repository);
            this.configsListProperty = new SimpleStringProperty(configsList.toString());
        }

        SimpleStringProperty getNameProperty() {
            return nameProperty;
        }

        SimpleStringProperty getRepositoryProperty() {
            return repositoryProperty;
        }

        SimpleStringProperty getLastConfigsListProperty() {
            return configsListProperty;
        }
    }

    public Button addPackageButton;
    public TreeTableView<PackageRow> packagesTree;
    public TreeTableColumn<PackageRow, String> packageNameColumn;
    public TreeTableColumn<PackageRow, String> repositoryColumn;
    public TreeTableColumn<PackageRow, String> configsListColumn;

    private void updatePackagesTable() {
        PackageService packageService = new PackageService(user.getID(), user.getPassword());
        List<Package> userPackages = packageService.getAll();

        TreeItem<PackageRow> root = new TreeItem<>(new PackageRow("", "", new ArrayList<>()));
        packageNameColumn.setCellValueFactory(
                (TreeTableColumn.CellDataFeatures<PackageRow, String> param)
                        -> param.getValue().getValue().getNameProperty()
        );
        repositoryColumn.setCellValueFactory(
                (TreeTableColumn.CellDataFeatures<PackageRow, String> param)
                        -> param.getValue().getValue().getRepositoryProperty()
        );
        configsListColumn.setCellValueFactory(
                (TreeTableColumn.CellDataFeatures<PackageRow, String> param)
                        -> param.getValue().getValue().getLastConfigsListProperty()
        );

        HashMap<String, ArrayList<PackageRow>> packagesByRepository = new HashMap<>();
        for (Package userPackage : userPackages) {
            if (packagesByRepository.containsKey(userPackage.getRepository().getName())) {
                packagesByRepository.get(userPackage.getRepository().getName()).add(
                        new PackageRow(userPackage.getName(), userPackage.getRepository().getName(), userPackage.getFiles())
                );
            } else {
                packagesByRepository.put(
                        userPackage.getRepository().getName(),
                        new ArrayList<>(
                                Collections.singletonList(
                                        new PackageRow(userPackage.getName(), userPackage.getRepository().getName(), userPackage.getFiles())
                                )
                        )
                );
            }
        }

        for (Map.Entry<String, ArrayList<PackageRow>> entry : packagesByRepository.entrySet()) {
            TreeItem<PackageRow> dir = new TreeItem<>(new PackageRow(entry.getKey(), "", new ArrayList<>()));
            for (PackageRow _package : entry.getValue()) {
                TreeItem<PackageRow> row = new TreeItem<>(_package);
                dir.getChildren().add(row);
            }
            root.getChildren().add(dir);
        }

        packagesTree.setRoot(root);
        packagesTree.setShowRoot(false);
    }

    public void addPackageClicked(ActionEvent actionEvent) {

    }

    private void insertPackageIntoDB() {

    }

    private void installPackage(Package pkg) {

    }
}

File: src/main/kotlin/ui/MainWindowController.kt
package ui

import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleStringProperty
import javafx.beans.value.ObservableValue
import javafx.fxml.FXMLLoader
import javafx.geometry.Pos
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.control.cell.CheckBoxTreeTableCell
import javafx.scene.layout.AnchorPane
import javafx.stage.Modality
import javafx.stage.Stage
import javafx.util.Callback
import models.File
import models.Package
import models.Repository
import models.User
import services.FileService
import services.PackageService
import services.RepositoryService
import ui.packages.AddPackageDialogController
import ui.packages.PackageConfigManagerController
import utils.FileUtil
import utils.UIUtil
import java.io.IOException
import java.sql.Timestamp
import java.util.*


class MainWindowController {

    private lateinit var user: User

    lateinit var rootAP: AnchorPane
    lateinit var usernameLabel: Label

    internal fun setUser(user: User) {
        this.user = user
        usernameLabel.text = "Files of " + user.name

        updateFilesTable()
        updatePackagesTable()
    }


    // files logic

    lateinit var addFileButton: Button
    lateinit var filesTree: TreeTableView<FileRow>
    lateinit var fileNameColumn: TreeTableColumn<FileRow, String>
    lateinit var createdColumn: TreeTableColumn<FileRow, String>
    lateinit var modifiedColumn: TreeTableColumn<FileRow, String>
    lateinit var fileDownloadedColumn: TreeTableColumn<FileRow, Boolean>

    inner class FileRow(var id: List<Int>, name: String, created: Timestamp, lastModified: Timestamp, downloaded: Boolean) {
        var nameProperty = SimpleStringProperty(name)
        var createdProperty = SimpleStringProperty(
                if (created.equals(Timestamp(0)))
                    ""
                else
                    created.toString()
        )
        var lastModifiedProperty = SimpleStringProperty(
                if (lastModified.equals(Timestamp(0)))
                    ""
                else
                    lastModified.toString()
        )
        var downloadedProperty = SimpleBooleanProperty(downloaded)
    }

    private fun updateFilesTable() {
        val fileService = FileService(user.id, user.password)
        val userFiles = fileService.getAll()

        val root = TreeItem(FileRow(listOf(-1), "n", Timestamp(0), Timestamp(0), true)) // TODO: check all the files in this directory

        fileNameColumn.cellFactory = Callback<TreeTableColumn<FileRow, String>, TreeTableCell<FileRow, String>> {
            object : TreeTableCell<FileRow, String>() {
                var btn = Button("X")

                public override fun updateItem(item: String?, empty: Boolean) {
                    super.updateItem(item, empty)
                    if (empty) {
                        graphic = null
                        text = null
                    } else {
                        btn.setOnAction {
                            for (id in treeTableRow.item.id) {
                                fileService.delete(fileService.find(id)!!)
                            }
                            updateFilesTable()
                            updatePackagesTable()
                        }
                        if (treeTableRow.item != null) {
                            graphic = btn
                            text = treeTableRow.item.nameProperty.get()
                        } else {
                            graphic = null
                            text = null
                        }
                    }
                }
            }
        }

        fileNameColumn.setCellValueFactory { param: TreeTableColumn.CellDataFeatures<FileRow, String> ->
            param.value.value.nameProperty
        }
        createdColumn.setCellValueFactory { param: TreeTableColumn.CellDataFeatures<FileRow, String> ->
            param.value.value.createdProperty
        }
        modifiedColumn.setCellValueFactory { param: TreeTableColumn.CellDataFeatures<FileRow, String> ->
            param.value.value.lastModifiedProperty
        }

        fileDownloadedColumn.cellValueFactory = Callback<TreeTableColumn.CellDataFeatures<FileRow, Boolean>, ObservableValue<Boolean>> { param ->
            val file = param.value.value
            val downloadedProperty = SimpleBooleanProperty(file.downloadedProperty.get())
            downloadedProperty.addListener { _, _, newValue ->
                downloadedProperty.set(newValue)
                when (newValue) {
                    true -> {
                        for (id in file.id) {
                            FileUtil.downloadFile(fileService.find(id)!!)
                        }
                    }
                    false -> {
                        for (id in file.id) {
                            FileUtil.deleteFile(fileService.find(id)!!)
                        }
                    }
                }
                updateFilesTable()
            }
            downloadedProperty
        }

        fileDownloadedColumn.cellFactory = Callback<TreeTableColumn<FileRow, Boolean>, TreeTableCell<FileRow, Boolean>> {
            val cell = CheckBoxTreeTableCell<FileRow, Boolean>()
            cell.alignment = Pos.CENTER
            cell.isEditable = true
            cell
        }
        filesTree.isEditable = true

        val filesByDir = HashMap<String, ArrayList<FileRow>>()
        for (userFile in userFiles) {
            val newRow = FileRow(
                    listOf(userFile.id),
                    userFile.name,
                    userFile.created!!,
                    userFile.modified!!,
                    FileUtil.fileIsDownloaded("${userFile.path}/${userFile.name}")
            )
            if (filesByDir.containsKey(userFile.path)) {
                filesByDir[userFile.path]!!.add(newRow)
            } else {
                filesByDir[userFile.path!!] = ArrayList(listOf(newRow))
            }
        }

        for ((key, value) in filesByDir) {
            val children = mutableListOf<TreeItem<FileRow>>()
            val ids = mutableListOf<Int>()
            var dirIsDownloaded = true
            for (_file in value) {
                val row = TreeItem(_file)
                ids.addAll(_file.id)
                children.add(row)
                if (!_file.downloadedProperty.get())
                    dirIsDownloaded = false
            }
            // TODO: add directory created and modified times
            val dir = TreeItem(FileRow(ids, key, Timestamp(0), Timestamp(0), dirIsDownloaded))
            dir.children.addAll(children)
            root.children.add(dir)
        }
        filesTree.root = root
        filesTree.isShowRoot = false
    }

    @Throws(IOException::class)
    fun addFileClicked() {
        UIUtil.addFile(FileService(user.id, user.password))
        updateFilesTable()
    }


    // packages logic

    lateinit var addPackageButton: Button
    lateinit var packagesTree: TreeTableView<PackageRow>
    lateinit var packageNameColumn: TreeTableColumn<PackageRow, String>
    lateinit var repositoryColumn: TreeTableColumn<PackageRow, String>
    lateinit var configsListColumn: TreeTableColumn<PackageRow, String>

    inner class PackageRow(var id: List<Int>, name: String, repository: String, configsList: List<File>, var isRepo: Boolean?) {
        var nameProperty = SimpleStringProperty(name)
        var repositoryProperty = SimpleStringProperty(repository)
        var configsListProperty = SimpleStringProperty(
                if (configsList.isEmpty())
                    ""
                else
                    run {
                        var result = "${configsList[0].path}/${configsList[0].name}"
                        if (configsList.size > 1) {
                            for (config in configsList.drop(1))
                                result += ", ${config.path}/${config.name}"
                        }
                        result
                    }
        )
    }

    private fun updatePackagesTable() {
        val packageService = PackageService(user.id, user.password)
        val userPackages = packageService.getAll()

        val root = TreeItem(PackageRow(listOf(-1), "", "", ArrayList(), true))

        packageNameColumn.cellFactory = Callback<TreeTableColumn<PackageRow, String>, TreeTableCell<PackageRow, String>> {
            object : TreeTableCell<PackageRow, String>() {
                var btn = Button("X")

                public override fun updateItem(item: String?, empty: Boolean) {
                    super.updateItem(item, empty)
                    if (empty) {
                        graphic = null
                        text = null
                    } else {
                        btn.setOnAction {
                            for (id in treeTableRow.item.id) {
                                packageService.delete(packageService.find(id)!!)
                            }
                            updatePackagesTable()
                            updateFilesTable()
                        }
                        if (treeTableRow.item != null) {
                            graphic = btn
                            text = treeTableRow.item.nameProperty.get()
                        } else {
                            graphic = null
                            text = null
                        }
                    }
                }
            }
        }

        packageNameColumn.setCellValueFactory { param: TreeTableColumn.CellDataFeatures<PackageRow, String> ->
            param.value.value.nameProperty
        }
        repositoryColumn.setCellValueFactory { param: TreeTableColumn.CellDataFeatures<PackageRow, String> ->
            param.value.value.repositoryProperty
        }
        configsListColumn.cellFactory = Callback<TreeTableColumn<PackageRow, String>, TreeTableCell<PackageRow, String>> {
            object : TreeTableCell<PackageRow, String>() {
                var btn = Button("Edit")

                public override fun updateItem(item: String?, empty: Boolean) {
                    super.updateItem(item, empty)
                    if (empty) {
                        graphic = null
                        text = null
                    } else {
                        btn.setOnAction {
                            manageConfigs(packageService.find(treeTableRow.item.id[0])!!) // there's a guarantee that this row can be only a package
                            updatePackagesTable()
                            updateFilesTable()
                        }
                        if (treeTableRow.item != null) {
                            if (!treeTableRow.item.isRepo!!) {
                                graphic = btn
                                text = treeTableRow.item.configsListProperty.get()
                            }
                        } else {
                            graphic = null
                            text = null
                        }
                    }
                }
            }
        }

        configsListColumn.setCellValueFactory { param: TreeTableColumn.CellDataFeatures<PackageRow, String> ->
            param.value.value.configsListProperty
        }

        val packagesByRepository = HashMap<String, ArrayList<PackageRow>>()
        for (userPackage in userPackages!!) {
            val newRow = PackageRow(
                    listOf(userPackage.id),
                    userPackage.name!!,
                    userPackage.repository!!.name,
                    userPackage.files!!,
                    false
            )
            if (packagesByRepository.containsKey(userPackage.repository!!.name)) {
                packagesByRepository[userPackage.repository!!.name]!!.add(newRow)
            } else {
                packagesByRepository[userPackage.repository!!.name] = ArrayList(listOf(newRow))
            }
        }

        for ((key, value) in packagesByRepository) {
            val children = mutableListOf<TreeItem<PackageRow>>()
            val ids = mutableListOf<Int>()
            for (_package in value) {
                val row = TreeItem(_package)
                children.add(row)
                ids.addAll(_package.id)
            }
            val repo = TreeItem(PackageRow(ids, key, "", ArrayList(), true))
            repo.children.addAll(children)
            root.children.add(repo)
        }

        packagesTree.root = root
        packagesTree.isShowRoot = false
    }

    private fun manageConfigs(pkg: Package) {
        val loader = FXMLLoader(javaClass.getResource("/PackageConfigManager.fxml"))
        val root = loader.load<Parent>()
        val controller = loader.getController<PackageConfigManagerController>()
        controller.setPackage(pkg, user)

        val stage = Stage()
        stage.scene = Scene(root)
        stage.initModality(Modality.APPLICATION_MODAL)
        stage.isResizable = false
        stage.showAndWait()
    }

    @Throws(Exception::class)
    fun addPackageClicked() {
        val repositoryService = RepositoryService(user.id, user.password)
        if (repositoryService.find(2).manager == "no_manager") { // on fresh installation
            repositoryService.update(Repository.default())
        }

        val loader = FXMLLoader(javaClass.getResource("/AddPackageDialog.fxml"))
        val root = loader.load<Parent>()
        val controller = loader.getController<AddPackageDialogController>()

        val stage = Stage()
        stage.scene = Scene(root)
        stage.initModality(Modality.APPLICATION_MODAL)
        stage.isResizable = false
        stage.showAndWait()

        if (controller.nameInput.text.isEmpty()) {
            return
        }

        val configs = ArrayList<File>()
        val filesList = controller.configsPathInput.text
        if (!filesList.isEmpty()) {
            val filesPath = filesList.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            for (_path in filesPath) {
                val path = _path.trim { it <= ' ' }
                val tempFile = java.io.File(path)
                if (!tempFile.exists()) {
                    UIUtil.showInvalidPathAlert(path)
                    return
                }

                val filesService = FileService(user.id, user.password)
                val name = path.substring(path.lastIndexOf('/') + 1)
                val location = path.substring(0, path.lastIndexOf('/'))
                val entry = filesService.find(name, location)
                if (entry != null) { // TODO: fix error while adding a package with already added config file
                    configs.add(entry)
                } else {
                    if (tempFile.isFile) {
                        configs.add(FileUtil.insertFileIntoDB(tempFile, path, Package.default(), FileService(user.id, user.password)))
                        updateFilesTable()
                    }

                    if (tempFile.isDirectory) {
                        configs.addAll(FileUtil.insertCatalogIntoDB(tempFile, Package.default(), FileService(user.id, user.password)))
                        updateFilesTable()
                    }
                }
            }
        }

        var repository = repositoryService.find(controller.repositoryInput.value.url)

        if (repository == null) {
            repositoryService.save(controller.repositoryInput.value)
            repository = repositoryService.find(controller.repositoryInput.value.url)
        }

        val pkg = Package(
                controller.nameInput.text.trim { it <= ' ' },
                repository!!
        )
        pkg.files = configs

        val packageService = PackageService(user.id, user.password)
        packageService.save(pkg)
        updatePackagesTable()

        val fileService = FileService(user.id, user.password)
        for (config in configs) {
            config.`package` = pkg
            fileService.update(config)
        }
    }
}
